### PR TITLE
Mutual information metric

### PR DESCRIPTION
**Pull request by nicolaspi**
# Description

Adds the `MutualInformation` metric.

Implementation based on the mutual information estimator described in
_Weihao Gao, Sreeram Kannan, Sewoong Oh, and Pramod Viswanath. Estimating
mutual information for discrete-continuous mixtures._
https://arxiv.org/pdf/1709.06212v3.pdf

The estimator works on any kind of distribution: discrete, continuous or a mix of both.

This implementation has a memory complexity of O(`buffer_size`x`compute_batch_size`) and a time complexity of O(`buffer_size`^2). They can be adjusted with the corresponding parameters. Smaller value of `buffer_size` increases the bias of the estimator while small value of `compute_batch_size` increases the computation time.

The effect of the parameter `buffer_size` is depicted in the following figures. They show the performance of the estimator when _X_ and _Y_ are drawn from a multivariate gaussian for different correlation coefficients. We can see that lower values increase the bias, but preserves the variance.

image

image

## Type of change

- [X] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)

# Checklist:

- [X] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [X] By running pre-commit hooks
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by nicolaspi** [comment]
Hi @bhack, can you have a look at this?

**Comment by seanpmorgan** [comment]
Thank you for your contribution. We sincerely apologize for any delay in reviewing, but TensorFlow Addons is transitioning to a minimal maintenance and release mode. New features will not be added to this repository. For more information, please see our public messaging on this decision:
[TensorFlow Addons Wind Down](https://github.com/tensorflow/addons/issues/2807)

Please consider sending feature requests / contributions to other repositories in the TF community with a similar charters to TFA:
[Keras](https://github.com/keras-team/keras)
[Keras-CV](https://github.com/keras-team/keras-cv)
[Keras-NLP](https://github.com/keras-team/keras-nlp)
